File: bun/src/day5/day5-2.ts
import { RuleMap } from './lib'

export const fixLine = (vals: string[], ruleMap: RuleMap) => {
	const newLine = [...vals]
	return newLine.sort((a, b) => ruleMap.isValidOrder(a, b) ? -1 : 1)
}

export const solve = (input: string) => {
	const [rulesStr, inputStr] = input.split('\n\n')
	const rules = rulesStr.split('\n')
	const ruleMap = RuleMap.fromRules(rules).ensureHasAllRelations()
	return inputStr.split('\n').reduce((sum, line) => {
		const vals = line.split(',')
		const rowIsOk = vals.every((testerVal, testerIdx) => {
			let testerValPositionOk = true
			for (let i = 0; i < vals.length; i++) {
				const v = vals[i]
				if (i < testerIdx) {
					if (testerValPositionOk)
						testerValPositionOk = ruleMap.isValidOrder(v, testerVal)
				} else if (testerIdx < i) {
					if (testerValPositionOk)
						testerValPositionOk = ruleMap.isValidOrder(testerVal, v)
				}
			}
			return testerValPositionOk
		})
		
		if (!rowIsOk) {
			const fixedLine = fixLine(vals, ruleMap)
			return sum + parseInt(fixedLine[(fixedLine.length - 1) / 2])
		} else {
			return sum
		}
	}, 0)
}

File: rust/day5/src/bin/day5-2.rs
use aoc::solve_input_file_to_int;
use day5::model::RuleMap;

pub fn fix_line<'a>(vals: Vec<&'a str>, rulemap: &RuleMap) -> Vec<&'a str> {
    let mut new_line = vals.clone();
    new_line.sort_by(|a, b| match rulemap.is_valid_order(a, b) {
        true => std::cmp::Ordering::Greater,
        false => std::cmp::Ordering::Less,
    });
    new_line
}

fn solve(input: String) -> u64 {
    let (rules_str, input_str) = input.split_once("\n\n").unwrap();
    let rules = rules_str.lines().collect::<Vec<&str>>();
    let rulemap = RuleMap::from_rules(rules).ensure_has_all_relations();

    // for every item in input_str
    input_str.lines().fold(0u64, |sum, line| {
        // println!("\ntesting {}", line);
        let vals = line.split(",").collect::<Vec<&str>>();
        match vals
            .clone()
            .iter()
            .enumerate()
            .all(|(tester_idx, tester_val)| {
                (0..vals.len() as usize).all(|idx| match vals[idx] {
                    v if idx < tester_idx => rulemap.is_valid_order(v, tester_val),
                    v if tester_idx < idx => rulemap.is_valid_order(tester_val, v),
                    _ => true,
                })
            }) {
            true => sum,
            _ => {
                let fixed_line = fix_line(vals, &rulemap);
                sum + fixed_line[(fixed_line.len() - 1) / 2]
                    .parse::<u64>()
                    .unwrap()
            }
        }
    })
}

fn main() {
    solve_input_file_to_int(solve)
}
